packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts

```diff
@@ -166,10 +166,16 @@ export class AggregatedAttestationPool {
       }
     }
 
-    return attestationsByScore
-      .sort((a, b) => b.score - a.score)
-      .slice(0, MAX_ATTESTATIONS)
-      .map((attestation) => attestation.attestation);
+    const sortedAttestationsByScore = attestationsByScore.sort((a, b) => b.score - a.score);
+    const attestationsForBlock: phase0.Attestation[] = [];
+    for (const [i, attestationWithScore] of sortedAttestationsByScore.entries()) {
+      if (i >= MAX_ATTESTATIONS) {
+        break;
+      }
+      // attestations could be modified in this op pool, so we need to clone for block
```

Comment:
Nice one! I'm wondering if this could be tested somehow?

Also immutable data structure (à la immutable.js) could be considered if the cloning is identified as a perf issue (here or elsewhere).

Reply:
> Nice one! I'm wondering if this could be tested somehow?

testing this requires setting up the pool and add attestations and produce blocks at the same time, I think it's not worth it. We do have a lot of unit tests for this file https://github.com/ChainSafe/lodestar/blob/unstable/packages/beacon-node/test/unit/chain/opPools/aggregatedAttestationPool.test.ts

> Also immutable data structure (à la immutable.js) could be considered if the cloning is identified as a perf issue (here or elsewhere).

we only need to do the clone in this specific scenario, we do use immutable.js in this PR https://github.com/ChainSafe/lodestar/pull/6042 where it's really needed

Reply:
Makes sense!

Also might be clearer to just add the cloning part in the existing `map`, as map already returns a new array anyway. A matter of taste probably.

Reply:
> Also might be clearer to just add the cloning part in the existing map

Yes I agree there. No need to be so much more verbose with named intermediate variables.

Reply:
`clone()` made the code a little bit slower, so I just want to use a single for loop to hopefully keep the same performance. There was already an issue with producing slow beacon block body #5793